File: apps/backend/inventory/products/products.service.go
package products

import (
	"backend/types"

	"github.com/labstack/echo/v4"
)

type ProductsService struct {
	Repository ProductsRepository
}

func NewProductsService() *ProductsService {

	return &ProductsService{
		Repository: NewProductsPostgreSQLRepository(),
	}
}

func (service ProductsService) GetPage(filter types.PageFilter[ProductPageFilter]) types.PageData[Product] {

	return service.Repository.GetPage(filter)
}

func (service ProductsService) Save(body SaveProductDto) error {

	newProduct := NewProduct(
		"1",
		body.Name,
	)
	service.Repository.Save(newProduct)

	return nil
}

func (service ProductsService) Update(id string, body SaveProductDto) error {

	product, err := service.Repository.GetById(id)

	if err != nil {
		return echo.ErrNotFound
	}

	service.Repository.Update(product)
	return nil

}

func (service ProductsService) Delete(id string) error {
	service.Repository.Delete(id)

	return nil
}

File: apps/penguin/src/products/products.service.ts
import {
  ConflictException,
  Inject,
  Injectable,
  NotFoundException,
} from '@nestjs/common';
import ProductRepository from './product-repository.interface';
import LocalProductRepository from './local-product.repository';
import { SaveProductDto } from './product.dto';
import Product from './product.class';
import { PageData, PageFilter } from '../shared/types/pagination';

@Injectable()
export default class ProductsService {
  constructor(
    @Inject(LocalProductRepository)
    private readonly repository: ProductRepository,
  ) {}

  async save(body: SaveProductDto): Promise<void> {
    const exists = await this.repository.getByName(body.name);

    if (exists) {
      throw new ConflictException('Product already exists');
    }

    const newProduct = new Product({
      id: new Date().getTime().toString(),
      ...body,
    });

    await this.repository.save(newProduct);

    return;
  }

  async update(id: string, body: SaveProductDto): Promise<void> {
    const current = await this.getById(id);

    if (current.name !== body.name) {
      const exists = await this.repository.getByName(body.name);
      if (exists) {
        throw new ConflictException('Product already exists');
      }
    }

    const updated = new Product({
      ...current,
      name: body.name,
      description: body.description,
    });

    await this.repository.update(updated);

    return;
  }

  async delete(id: string): Promise<void> {
    await this.repository.delete(id);
    return;
  }

  async getById(id: string): Promise<Product> {
    const result = await this.repository.getById(id);

    if (!result) {
      throw new NotFoundException('Product not found');
    }

    return result;
  }

  async getPage(filter: PageFilter<object>): Promise<PageData<Product>> {
    return await this.repository.getPage(filter);
  }
}
